src/couch_epi_data.erl

```diff
@@ -12,131 +12,103 @@
-    {noreply, State}.
-
-handle_info(_Info, State) ->
-    {noreply, State}.
-
-terminate(_Reason, _State) ->
-    ok.
-
-code_change(_OldVsn, State, _Extra) ->
-    {_, NewState} = reload_if_updated(State),
-    {ok, NewState}.
+data(Specs) ->
+    Locators = locate_sources(Specs),
+    case lists:foldl(fun collect_data/2, {ok, [], []}, Locators) of
+        {ok, Hashes, Data} ->
+            {ok, couch_epi_util:hash(Hashes), Data};
+        Error ->
+            Error
+    end.
 
 %% ------------------------------------------------------------------
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
 
-reload_if_updated(#state{hash = OldHash, module = Module} = State) ->
-    case couch_epi_functions_gen:hash([Module]) of
-        OldHash ->
-            {ok, State};
-        Hash ->
-            safe_set(Hash, State)
+collect_data({App, Locator}, {ok, HashAcc, DataAcc}) ->
+    case definitions(Locator) of
+        {ok, Hash, Data} ->
+            {ok, [Hash | HashAcc], [{App, Data} | DataAcc]};
+        Error ->
+            Error
+    end;
+collect_data({_App, _Locator}, Error) ->
+    Error.
+
+extract_minimal_interval(Specs) ->
+    lists:foldl(fun minimal_interval/2, undefined, Specs).
+
+minimal_interval({_App, #couch_epi_spec{options = Options}}, Min) ->
+    case lists:keyfind(interval, 1, Options) of
+        {interval, Interval} -> min(Interval, Min);
+        false -> Min
     end.
 
-safe_set(Hash, #state{} = State) ->
-    #state{
-        handle = Handle,
-        subscriber = Subscriber,
-        module = Module,
-        key = Key} = State,
-    try
-        Data = get_from_module(Module),
-        OldData = couch_epi_data_gen:current_data(Handle, Subscriber),
-        ok = couch_epi_data_gen:set(Handle, Subscriber, Data),
-        couch_epi_server:notify(Subscriber, Key, {data, OldData}, {data, Data}),
-        {ok, State#state{hash = Hash}}
-    catch Class:Reason ->
-        {{Class, Reason}, State}
+locate_sources(Specs) ->
+    lists:map(fun({ProviderApp, #couch_epi_spec{value = Src}}) ->
+        {ok, Locator} = locate(ProviderApp, Src),
+        {ProviderApp, Locator}
+    end, Specs).
+
+locate(App, {priv_file, FileName}) ->
+    case priv_path(App, FileName) of
+        {ok, FilePath} ->
+            ok = ensure_exists(FilePath),
+            {ok, {file, FilePath}};
+        Else ->
+            Else
+    end;
+locate(_App, {file, FilePath}) ->
+    ok = ensure_exists(FilePath),
+    {ok, {file, FilePath}};
+locate(_App, Locator) ->
+    {ok, Locator}.
+
+priv_path(AppName, FileName) ->
+    case code:priv_dir(AppName) of
+        {error, _Error} = Error ->
+            Error;
+        Dir ->
+            {ok, filename:join(Dir, FileName)}
     end.
 
-get_from_module(Module) ->
-    Module:data().
+ensure_exists(FilePath) ->
+    case filelib:is_regular(FilePath) of
+        true ->
+            ok;
+        false ->
+            {error, {notfound, FilePath}}
```

Comment:
this function does not ensure the FilePath exists, so either the name or implementation is wrong.